### ISSUE TITLE
Decoding and encoding are not inverses for Text in structure.

### ISSUE THREAD

**Issue by NorfairKing** [bos/aeson#508]
Here is a custom data type:
``` Haskell
data ForShow =
    ForShow Text
    deriving (Show, Eq, Generic)
instance ToJSON ForShow
instance FromJSON ForShow
```
Here is the error:
```
  src/Test/Validity/Aeson.hs:85: 
  1) Test.Validity.Aeson, JSON ForShow (valid), decode :: ForShow -> Data.ByteString.Lazy.ByteString, ensures that encode and decode are inverses for "valid ForShow"'s
       Falsifiable (after 3 tests): 
       expected: Right (ForShow "\NUL\NUL")
        but got: Right (ForShow "\STX\NUL")
       ForShow "\NUL\NUL"
```

Please put some quickcheck properties in the test suite to ensure that this cannot happen again.

**Comment by NorfairKing** [comment]
Uh, Was testing on aeson-0.11. Currently testing on aeson-1.1.0.0.

**Comment by NorfairKing** [comment]
aeson-1.1.0.0 has a similar problem:

```
  src/Test/Validity/Aeson.hs:85: 
  1) Test.Validity.Aeson, JSON ForShow (valid), decode :: ForShow -> Data.ByteString.Lazy.ByteString, ensures that encode and decode are inverses for "valid ForShow"'s
       Falsifiable (after 1 test): 
       expected: Right (ForShow "\248")
        but got: Right (ForShow "\DC1")
       ForShow "\248"
```

**Comment by phadej** [comment]
I cannot reproduce:

```haskell
>>> :set -XDeriveGeneric -XOverloadedStrings 
>>> data ForShow = ForShow Text deriving (Show, Eq, Generic)
>>> instance FromJSON ForShow
>>> instance ToJSON ForShow
>>> decode (Data.Aeson.encode (ForShow "\248")) :: Maybe ForShow
Just (ForShow "\248")
```

with current master and using `stack-lts7.yaml`.

**Comment by NorfairKing** [comment]
I also cannot reproduce it in GHCI, but only in my test suite. This is very weird. Maybe it has something to do with OverloadedStrings?

Here is a failing test with `hspec` and `genvalidity-hspec-aeson`:

```
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

import Test.Hspec

import Data.Aeson
import Data.GenValidity
import Data.GenValidity.Text ()
import Data.Text (Text)
import GHC.Generics
import Test.Validity.Aeson


spec :: Spec
spec = jsonSpecOnValid @ForShow
data ForShow =
    ForShow Text
    deriving (Show, Eq, Generic)
instance Validity ForShow
instance GenUnchecked ForShow
instance GenValid ForShow
instance FromJSON ForShow
instance ToJSON ForShow
```

In ghci: 

```
stack ghci --package aeson-1.1.0.0      
Configuring GHCi with the following packages: 
GHCi, version 8.0.1: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /tmp/ghci26644/ghci-script
Prelude> :set -XDeriveGeneric -XOverloadedStrings 
Prelude> import Data.Text (Text)
Prelude Data.Text> import GHC.Generics
Prelude Data.Text GHC.Generics> import Data.Aeson
Prelude Data.Text GHC.Generics Data.Aeson> data ForShow = ForShow Text deriving (Show, Eq, Generic)
Prelude Data.Text GHC.Generics Data.Aeson> instance FromJSON ForShow
Prelude Data.Text GHC.Generics Data.Aeson> instance ToJSON ForShow
Prelude Data.Text GHC.Generics Data.Aeson> decode (Data.Aeson.encode (ForShow "\STX\NUL\29328")) :: Maybe ForShow
Just (ForShow "\STX\NUL\29328")
Prelude Data.Text GHC.Generics Data.Aeson> decode (Data.Aeson.encode (ForShow "\248")) :: Maybe ForShow
Just (ForShow "\248")
```

In the test suite:

```
  src/Test/Validity/Aeson.hs:85: 
  1) Test.Validity.Aeson, JSON ForShow (valid), decode :: ForShow -> Data.ByteString.Lazy.ByteString, ensures that encode and decode are inverses for "valid ForShow"'s
       Falsifiable (after 1 test): 
       expected: Right (ForShow "\248")
        but got: Right (ForShow "I")
       ForShow "\248"
```

**Comment by phadej** [comment]
Can you test with explicit `encode . decode` in test-suite, and not using `jsonSpecOnValid`?

**Comment by NorfairKing** [comment]
I think I don't understand what you mean by that.
Are you suggesting GHC is at fault?

For the record, here is the implementation of `jsonSpecOnValid`. 
It does a `decode . encode`, not `encode . decode`.

https://github.com/NorfairKing/validity/blob/master/genvalidity-hspec-aeson/src/Test/Validity/Aeson.hs

**Comment by phadej** [comment]
sorry, `encode .decode`. I mean: I'd like to see a test-case without relying on `jsonSpecOnValid`. The issue may be in `validity`. Also it would be great if `genvalidity-hspec-aeson` outputted the intermediate `ByteString`, as `Word8`. Maybe there's something weird with the `Text` value generated in the first place.

**Comment by NorfairKing** [comment]
That is a great suggestion for `genvalidity-hspec-aeson`!
I tried it out here: https://github.com/NorfairKing/validity/blob/aeson-failure/genvalidity-hspec-aeson/src/Test/Validity/Aeson.hs

Here are some of the outputs:

```
  src/Test/Validity/Aeson.hs:87: 
  1) Test.Validity.Aeson, JSON ForShow (valid), decode :: ForShow -> Data.ByteString.Lazy.ByteString, ensures that encode and decode are inverses for "valid ForShow"'s
       Falsifiable (after 3 tests): 
       Decoding succeeded, but the decoded value
       ForShow "\NUL\9648n"
       differs from expected decoded value
       ForShow "\NUL\n\NUL"
       'encode' encoded it to the json
       "\"\\u0000\226\150\176n\""
       , which is the following in [Word8]:
       [34,92,117,48,48,48,48,226,150,176,110,34]
       ForShow "\NUL\47273\254"


  src/Test/Validity/Aeson.hs:87: 
  1) Test.Validity.Aeson, JSON ForShow (valid), decode :: ForShow -> Data.ByteString.Lazy.ByteString, ensures that encode and decode are inverses for "valid ForShow"'s
       Falsifiable (after 1 test): 
       Decoding succeeded, but the decoded value
       ForShow "2"
       differs from expected decoded value
       ForShow "\205"
       'encode' encoded it to the json
       "\"2\""
       , which is the following in [Word8]:
       [34,50,34]
       ForShow "\205"


  src/Test/Validity/Aeson.hs:87: 
  1) Test.Validity.Aeson, JSON ForShow (valid), decode :: ForShow -> Data.ByteString.Lazy.ByteString, ensures that encode and decode are inverses for "valid ForShow"'s
       Falsifiable (after 5 tests): 
       Decoding succeeded, but the decoded value
       ForShow "\NUL"
       differs from expected decoded value
       ForShow "\207"
       'encode' encoded it to the json
       "\"\\u0000\""
       , which is the following in [Word8]:
       [34,92,117,48,48,48,48,34]
       ForShow "\207"
```

For the record, this is how valid `Text` values are generated:
https://github.com/NorfairKing/validity/blob/aeson-failure/genvalidity-text/src/Data/GenValidity/Text.hs#L29

So we find these problems, but then when I try to to add the failing tests in my own test suite, they don't fail.

``` 
    it "mytest" $ encodeAndDecodeAreInversesOnGen (pure $ ForShow "\205")
```

This just passes.

So I'm thinking the `genValid` somehow generates values that are valid according to `isValid` but not truly valid somehow, such that they are decoded incorrectly when they are encoded.

The validity instance is written here: https://github.com/NorfairKing/validity/blob/aeson-failure/validity-text/src/Data/Validity/Text.hs
and the default implementation of `genValid` is `genUnchecked suchThat isValid`.
When I implement `GenValid ForShow` with `genValid = ForShow <$> genValid`, everything works.
This makes me believe that `isValid` is not correct.

TLDR; I made a mistake, not aeson. I'm very sorry to have bothered you!
